### DISCUSSION TITLE
Synchronizing atoms with external sources

### DISCUSSION THREAD

**Discussion by mtbaqer** [pmndrs/jotai#1487] [Ideas]
This proposal is inspired by [recoil-sync](https://recoiljs.org/docs/recoil-sync/introduction).

I wanted to open discussion for a potential Util API for an easier way to synchronize atom values with external sources.

Of course we have `atomWithQuery` and `onMount` to help us do that currently, but I'm hoping we can push it even further.

Ideally the closer it is to how normally use atoms the better. One way of doing something close is by using `atomWithStorage` with the external source as an `AsyncStorage`.

Firebase example:

```tsx
import { get, getDatabase, onValue, ref, set } from "firebase/database";

const db = getDatabase();

const createFirebaseStorage = (): AsyncStorage<any> => ({
  getItem: (key) => get(ref(db, key)).then((snapshot) => snapshot.val()),
  setItem: (key, newValue) => set(ref(db, key), newValue),
  subscribe: (key, setAtom) => onValue(ref(db, key), (snapshot) => setAtom(snapshot.val())),
  removeItem: async () => {},
});
const firebaseStorage = createFirebaseStorage();

const atomWithFirebase = <T,>(path: string, defaultValue: T) => atomWithStorage<T>(path, defaultValue, firebaseStorage);

const countAtom = atomWithFirebase("count", 0);

function App() {
  const [count, setCount] = useAtom(countAtom);
  return (
    <>
      <button onClick={() => setCount((count) => count + 1)}> + </button>
      <h1>{count}</h1>
    </>
  );
}
```

Here we go beyond using `atomWithQuery` by having `setCount` also mutating the external state.

However, `atomWithStorage` seems to be pushed as more of a local persistence util than a synchronization one, don't really like how we need to make util functions to use the storages like `atomWithFirebase` above, and does not allow syncing with multiple sources at the same time.

I kind of like how `recoil-sync` is doing it, although not too keen on the current API, which using a wrapper to handle the synchronization logic.

The firebase example using the same idea would something like:
```tsx
//Root
const Root = () => (
  <SyncProvider
    getItem={(key) => get(ref(db, key)).then((snapshot) => snapshot.val())}
    setItem={(key, newValue) => set(ref(db, key), newValue)}
    subscribe={(key, setAtom) => onValue(ref(db, key), (snapshot) => setAtom(snapshot.val()))}
  >
    <App />
  </SyncProvider>
);

//Down the tree
const countAtom = atomWithSync("count", 0);

function Component() {
  const [count, setCount] = useAtom(countAtom);
  return (
    <>
      <button onClick={() => setCount((count) => count + 1)}>+</button>
      <h1>{count}</h1>
    </>
  );
}
```

One thing I'm not sure about is syncing with multiple sources, `recoil-sync` goes around this by using `storeKey` both in the provider and the atoms. I'm not sure how I feel about that.

**Comment by dai-shi** [comment]
Hi, thanks for opening up a discussion.

1st option: use `atomWithStorage`, which supports `subscribe` property in storage. I think firebase is a storage. There might be something missing, which should be improved.
2nd option: use `atomWithObservable`
3rd option: create a new `atomWithSync`, not sure how this is different from the above two. Maybe it should be a 3rd-party library. I don't prefer having non-optional Provider anyways.

**Comment by mtbaqer** [comment]
Thanks for the quick reply.

I think the emphasis here is not the `subscribe` functionality (a needed feature nonetheless), but instead having the getting and setting of external values be done through the conventional `[value, setValue]` pair.

The 1st option makes the most sense to me, so basically like the example above. I was just under the impression that `atomWithStorage` is mainly a local persistence util since its current description in docs is:
> The atomWithStorage function creates an atom with a value persisted in localStorage or sessionStorage for React or AsyncStorage for React Native.

So maybe we can add examples of using external storages? Maybe also rephrase the atom description above to mention that use case as well.

If we wanna go that route, is the example above the recommended pattern?

1. Make a `createStorage()` function
2. Use that function to make the storage
3. Make a `atomWithExternalStorage()` function to use everywhere

**Reply by dai-shi** [reply]
What's atomWithExternalStorage? Isn't it just atomWithStorage?

Can you elaborate your suggestion? If you say, this is a doc issue of atomWithStorage, I'd agree.

**Reply by mtbaqer** [reply]
Apologies for not being clear. Yes, I think improving the docs seems to be the course of action.

To elaborate more on what I meant, I was referring to these 3 calls:
```ts
const createFirebaseStorage = (): AsyncStorage<any> => ({...});
const firebaseStorage = createFirebaseStorage();
const atomWithFirebase = <T,>(path: string, defaultValue: T) => atomWithStorage<T>(path, defaultValue, firebaseStorage);
```
`atomWithExternalStorage()` is not the literal name but a user-made util function, in this case`atomWithFirebase()`, to use the storage they made.

Would this be the recommended pattern? 
The first two calls seem like they could be merged together using a util function similar to `createJSONStorage()` that we can add, any thoughts?

**Reply by dai-shi** [reply]
> Apologies for not being clear. Yes, I think improving the docs seems to be the course of action.
> `atomWithExternalStorage()` is not the literal name but a user-made util function, in this case`atomWithFirebase()`, to use the storage they made.

Yeah, that looks good.

btw, `createJSONStorage()` is bad at typing, so ideally, we should have a type guard included. That's also the point we should cover to improve the docs.

Sounds like you have a good perception to improve docs. Feel free to open a PR for docs. Follow-up questions are welcome.

**Reply by mtbaqer** [reply]
Can you elaborate more on the `createJSONStorage()` type guard?

And on the same note, what are your thoughts on making a similar function to simplify the pattern above? Maybe call it `createStorage()`?

Would be more than happy open PR for both as well.

**Reply by dai-shi** [reply]
createJSONStorage is a non-ideal implementation because it simply uses JSON.parse, whose return type is `any`.
For type safe usage, I highly discourage to use createJSONStorage. And, it's not clearly documented...

I'm not sure if `createStorage()` is necessary. How does it look like? Feels like it's a field for 3rd party library.
